File: examples/js/error_handling.js
/*
 * file: error_handling.js
 * description: demonstrates how to detect syntax errors and print error offsets
 * using the cjsonx wasm engine in javascript.
 */

const cjsonx = require('../../js/cjsonx.js');

async function main() {
    await cjsonx.ready;

    // malformed json string
    const brokenJson = '{"name": "cjsonx", "valid": false, }'; // trailing comma

    console.log("attempting to parse malformed json...\n");

    const success = cjsonx.parse(brokenJson);

    if (!success) {
        const errorMsg = cjsonx.getError();
        const errorOffset = cjsonx.getErrorOffset();
        
        console.log("error detected!");
        console.log(`message: ${errorMsg}`);
        console.log(`offset:  ${errorOffset}\n`);

        // point to the error
        console.log("json snippet:");
        console.log(brokenJson);
        let pointer = "";
        for (let i = 0; i < errorOffset; i++) pointer += " ";
        pointer += "^";
        console.log(pointer);
    } else {
        console.log("parsed successfully");
        cjsonx.free();
    }
}

main().catch(console.error);

File: examples/python/error_handling.py
"""
file: error_handling.py
description: demonstrates how to catch syntax errors and use doc.error_offset
to locate the exact position of the error in python.
"""

import cjsonx

def main():
    # a malformed json string (missing closing quote)
    json_data = '{\n  "name": "cjsonx,\n  "fast": true\n}'

    print("attempting to parse malformed json...\n")

    # 1. parse the json string
    doc = cjsonx.parse(json_data)

    # 2. check if valid
    if not doc.is_valid:
        print("error detected!")
        print(f"message: {doc.error}")
        print(f"offset:  {doc.error_offset}\n")

        # 3. visually point to the error
        start = max(0, doc.error_offset - 15)
        snippet = json_data[start:doc.error_offset + 10].replace('\n', ' ')
        print("json snippet:")
        print(f"{snippet}...")
        spaces = doc.error_offset - start
        print(" " * spaces + "^")
    else:
        print("parsed successfully (this should not happen for this input)")

if __name__ == "__main__":
    main()
